src/test/test_nodelist.c

```diff
@@ -1246,6 +1247,136 @@ test_nodelist_router_get_verbose_nickname(void *arg)
   return;
 }
 
+static void
+test_nodelist_routerstatus_has_changed(void *arg)
+{
+  (void)arg;
+  routerstatus_t rs_orig, rs;
+  memset(&rs_orig, 0, sizeof(rs_orig));
+  strlcpy(rs_orig.nickname, "friendly", sizeof(rs_orig.nickname));
+  memcpy(rs_orig.identity_digest, "abcdefghijklmnopqrst", 20);
+  memcpy(rs_orig.descriptor_digest, "abcdefghijklmnopqrst", 20);
+  rs_orig.addr = 0x7f000001;
+  rs_orig.or_port = 3;
+  rs_orig.published_on = time(NULL);
+
+#define COPY() memcpy(&rs, &rs_orig, sizeof(rs))
+#define ASSERT_SAME() \
+  STMT_BEGIN                                                    \
+    tt_assert(! routerstatus_has_changed(&rs_orig, &rs));       \
+    COPY();                                                     \
+  STMT_END
+#define ASSERT_CHANGED() \
+  STMT_BEGIN                                                    \
+    tt_assert(routerstatus_has_changed(&rs_orig, &rs));         \
+    COPY();                                                     \
+  STMT_END
+
+  COPY();
+  ASSERT_SAME();
+
+  rs.addr = 0x7f000002;
+  ASSERT_CHANGED();
+
+  strlcpy(rs.descriptor_digest, "hello world", sizeof(rs.descriptor_digest));
+  ASSERT_CHANGED();
+
+  strlcpy(rs.nickname, "fr1end1y", sizeof(rs.nickname));
+  ASSERT_CHANGED();
+
+  rs.published_on += 3600;
+  ASSERT_CHANGED();
+
+  rs.or_port = 55;
+  ASSERT_CHANGED();
+
+  rs.dir_port = 9999;
+  ASSERT_CHANGED();
+
+  tor_addr_parse(&rs.ipv6_addr, "1234::56");
+  ASSERT_CHANGED();
+
+  rs.ipv6_orport = 22;
+  ASSERT_CHANGED();
+
+  rs.is_authority = 1;
+  ASSERT_CHANGED();
+
+  rs.is_exit = 1;
+  ASSERT_CHANGED();
+
+  rs.is_stable = 1;
+  ASSERT_CHANGED();
+
+  rs.is_fast = 1;
+  ASSERT_CHANGED();
+
+  rs.is_flagged_running = 1;
+  ASSERT_CHANGED();
+
+  // Isn't this obsolete?
+  rs.is_named = 1;
+  ASSERT_CHANGED();
+
+  // Isn't this obsolete?
+  rs.is_unnamed = 1;
+  ASSERT_CHANGED();
+
+  rs.is_valid = 1;
+  ASSERT_CHANGED();
+
+  rs.is_possible_guard = 1;
+  ASSERT_CHANGED();
+
+  rs.is_bad_exit = 1;
+  ASSERT_CHANGED();
+
+  rs.is_hs_dir = 1;
+  ASSERT_CHANGED();
+
+  rs.is_v2_dir = 1;
+  ASSERT_CHANGED();
+
+  rs.is_staledesc = 1;
+  ASSERT_CHANGED();
+
+  rs.has_bandwidth = 1;
+  ASSERT_CHANGED();
+
+  // Does not actually matter unless exitsummary changes.
```

Comment:
It actually doesn't matter at all, exitsummary is not output via the control port routerstatus format. (See exitsummary below for details.)

routerstatus_has_changed() Is only checking for changes that get output over the control port:
> It only checks for fields that are output by control port.
> This should be kept in sync with the struct routerstatus_t
> and the printing function routerstatus_format_entry in
> NS_CONTROL_PORT mode.

Let's rename routerstatus_has_changed() to a name that reflects its purpose as a control-port only function.

Reply:
In fact, has_exitsummary is not used in any routerstatus format. Perhaps it's used in microdescriptors?